✨ Upgrade cert-manager

munnerz opened the pull request:
**What this PR does / why we need it**:

Upgrades cert-manager to ~v0.16.0-alpha.0~ ~v0.15.2~ v0.16.0.

~I can adjust this PR to install the latest *stable* version if preferred.~ Done 🎉 

The new 'wait for ready' logic should work more universally by actually attempting to create an Issuer and Certificate resource as-per the documentation we provide in the installation guide: https://cert-manager.io/docs/installation/kubernetes/#verifying-the-installation

**Please note**: This PR does *not* handle the *upgrade* process for cert-manager if a user already has it installed in their cluster (either through `clusterctl init` or through any other means).

**Which issue(s) this PR fixes**:
Fixes #2240

vincepri commented:
/hold 

To wait for v0.3.7 to be released first

munnerz commented:
`apidiff` failing due to:

```
  Incompatible changes:
  - CertManagerClient.EnsureCertManager: added
  - CertManagerClient.EnsureWebhook: removed
```

I can keep the old method name around if preferred, and maybe introduce the new name as an extra function and deprecated the old one? `EnsureWebhook` isn't really very descriptive given the method actually installs *cert-manager*, not just the webhook :)

vincepri commented:
The job is option, I should have said something earlier. AFAIK these methods are only used within clusterctl, so it should be fine to change

vincepri commented:
/milestone v0.3.x

wfernandes commented on cmd/clusterctl/client/cluster/cert_manager.go:
```diff
@@ -94,29 +87,20 @@ func (cm *certManagerClient) Images() ([]string, error) {
 }
 
-// EnsureWebhook makes sure the cert-manager Web-hook is Available in a cluster:
+// EnsureCertManager makes sure the cert-manager Web-hook is Available in a cluster:
```
I suggest removing the "webhook is available" part since we are no longer doing that.

```suggestion
// EnsureCertManager ensures cert-manager is installed correctly by creating some test certs in the `cert-manager-test` namespace.
```

munnerz replied:
Done :)

wfernandes commented on cmd/clusterctl/client/cluster/cert_manager.go:
```diff
@@ -44,9 +45,9 @@ const (
 // CertManagerClient has methods to work with cert-manager components in the cluster.
 type CertManagerClient interface {
-	// EnsureWebhook makes sure the cert-manager Webhook is Available in a cluster:
+	// EnsureCertManager makes sure the cert-manager Webhook is Available in a cluster:
```
Maybe we update the entire comment since the Webhook here was referring to the API service webhook.

munnerz commented on cmd/clusterctl/client/cluster/cert_manager.go:
```diff
@@ -94,29 +87,20 @@ func (cm *certManagerClient) Images() ([]string, error) {
+func (cm *certManagerClient) EnsureCertManager() error {
 	log := logf.Log
 
-	// Checks if the cert-manager web-hook already exists, if yes, exit immediately
```
I've re-added this check with an additional function that calls Get on each resource in the cluster to verify whether cert-manager resources already exist before Creating them all again.

Now that I think some more though, it _may_ be better to just call Create(), because currently if a single IsNotFound error is encountered, we go ahead and call Create anyway.

Whilst we may not want to address a proper upgrade process in this PR, because it _is_ upgrading the manifest used, we probably want to implement some guards to avoid users downloading the new `clusterctl` and running `init` against an existing management cluster, as they could end out with orphaned resources from the previous release leftover.

I'm actually not 100% certain on how best to address this - updating the manifests in place *without* a defined upgrade process is tough, because we don't have any means to understand whether an older version _is_ installed (unless we retain both copies of the cert-manager install manifests and explicitly check if all resources in an older version exist, but that gets.. slippery).

We _do_ have this annotation on resources: `helm.sh/chart: 'cert-manager-v0.15.2'` but I'm not sure if relying on the cert-manager project to always set this in this format is wise. Perhaps we could annotate created objects (similar to what's done already), but with a version number too? For the 'transitionary period' for users coming from earlier clusterctl versions, we could check for the presence of the `cert-manager` value with _no_ version annotation in the `clusterctl.cluster.x-k8s.io/core` annotation). But without recording exactly what resources are required to be installed in earlier versions of cert-manager, it won't be possible for us to be certain *all* resources exist (i.e. if a previous invocation of `clusterctl init` failed half way through, we'll struggle to recover).

fabriziopandini commented on cmd/clusterctl/client/cluster/cert_manager.go:
```diff
@@ -94,32 +91,34 @@ func (cm *certManagerClient) Images() ([]string, error) {
-	// Gets the cert-manager objects from the embedded assets.
-	objs, err := cm.getManifestObjs()
+	// Skip re-installing cert-manager if all objects already exist
+	allObjectsExist, err := cm.allObjectsExist(ctx, log, objs)
```
I'm still wrapping up my mind around this check.
What If I created a cluster with an older version of clusterctl, installing cert-manager v0.11.0, what happens if I run clusterctl init with a new version of clusterctl (running init multiple times is a supported use case, eg. for installing an additional provider).
If I got this right, this will detect some object is missing, and it will install v0.15.2 on top of v0.11.0 without proper cleanup. This makes init get into the business of doing upgrades and I would like to avoid this.

What about instead running only waitForAPIReady to detect cert-manager is there? are there counter-indications?

munnerz replied:
Yeah, I added this check just this morning, but it occurs to me that you're right and it won't work: https://github.com/kubernetes-sigs/cluster-api/pull/3313#discussion_r452082535

+1 to just running `waitForAPIReady`, although we'll need to do it with a 0 timeout or something (i.e. just check once)

fabriziopandini replied:
second tough, this can lead to a deadlock if something goes wrong in the middle of an installation.
allObjectsExist == false, waitForAPIReady == false

fabriziopandini replied:
last bit. eventually, if it required to check if something is already installed, all the cert-manager components are labeled by clusterctl

munnerz replied:
Updated to no longer use `allObjectsExist` - now it just uses `waitForAPIReady`, and if that fails after one attempt, will call `createObj`.

munnerz replied:
resolving this thread as we no longer have this code 😄

fabriziopandini commented on cmd/clusterctl/client/cluster/cert_manager.go:
```diff
@@ -220,86 +218,79 @@ func (cm *certManagerClient) createObj(o unstructured.Unstructured) error {
-// getWebhook returns the cert-manager Webhook or nil if it does not exists.
-func (cm *certManagerClient) getWebhook() (*unstructured.Unstructured, error) {
+// allObjectsExist checks if all the objects with the given name already exist
+func (cm *certManagerClient) allObjectsExist(ctx context.Context, log logr.Logger, objs []unstructured.Unstructured) (bool, error) {
```
nit
```suggestion
func (cm *certManagerClient) allObjectsExist(ctx context.Context, objs []unstructured.Unstructured) (bool, error) {
```
We usually avoid to pass the logger around and instead use `log := logf.Log` to access to the global logger; same apply to other func

fabriziopandini commented on cmd/clusterctl/client/cluster/cert_manager.go:
```diff
@@ -220,86 +218,79 @@ func (cm *certManagerClient) createObj(o unstructured.Unstructured) error {
-		conditionStatus, ok := conditionMap["status"]
-		if !ok {
-			return false, errors.Wrapf(err, "invalid cert-manager web-hook: there %q condition does not have the status field", "Available")
+		cl, err := cm.proxy.NewClient()
```
nit.
what about creating cm.deleteObj (for simmetry with cm.createObj)

munnerz replied:
Can do - but it seems `createObj` exists in order to handle adding the label, whereas Delete does not have this requirement. Happy to make the change, but it'll basically be a one-liner function :)

fabriziopandini replied:
we can move inside the function the creation of the client...

munnerz replied:
Do we want to create a new client for each call to `createObj` and `deleteObj`? Or pass it in as an argument?

fabriziopandini replied:
we can move inside the function the creation of the client...
the client instance is cached. so fine to create 1 per call and keep the method signature clean

munnerz replied:
done

fabriziopandini commented on cmd/clusterctl/client/cluster/cert_manager.go:
```diff
@@ -220,86 +218,79 @@ func (cm *certManagerClient) createObj(o unstructured.Unstructured) error {
+// Issuer and Certificate).
+// This ensures that the Kubernetes apiserver is ready to serve resources within the
+// cert-manager API group.
+func (cm *certManagerClient) waitForAPIReady(ctx context.Context, log logr.Logger) error {
```
Just for confirmation, in this func we are creating an issurer/a certificate, but we are not waiting for the certificate to be processed as documented https://cert-manager.io/docs/installation/kubernetes/#verifying-the-installation.
Is this enough for assuming cert-manager is working?

fabriziopandini replied:
Another question, as far as I know cert-manager, this test can work both with v0.11.0 and v0.15.2

munnerz replied:
> Is this enough for assuming cert-manager is working?

So it depends on how you want to handle this - you are correct, that it may not yet be ready to issue certificates as the controller could still be leader electing. However, presumably the parts of code that create the Certificate resource also then wait for those Certificates to be *Ready* before consuming the signed certificate/private key. Would you rather wait at the start for the controller to be ready, or wait until further into execution? Easy enough to change either way 😄 

> Another question, as far as I know cert-manager, this test can work both with v0.11.0 and v0.15.2

Correct, this is a far more universal test - if this stops working, it'll be because:

1) The v1alpha2 API version no longer exists in cert-manager
2) We made a big mistake in cert-manager 😅

fabriziopandini commented on cmd/clusterctl/client/cluster/cert_manager.go:
```diff
@@ -220,86 +218,79 @@ func (cm *certManagerClient) createObj(o unstructured.Unstructured) error {
-// This is required to check the web-hook is working and ready to accept requests.
-func (cm *certManagerClient) isWebhookAvailable(webhook *unstructured.Unstructured) (bool, error) {
-	conditions, found, err := unstructured.NestedSlice(webhook.Object, "status", "conditions")
+	testObjs, err := getTestResourcesManifestObjs()
```
Ideally, we should randomize the namespace name, but this can be addressed in a follow up as well

fabriziopandini commented on cmd/clusterctl/client/cluster/cert_manager.go:
```diff
@@ -94,32 +91,34 @@ func (cm *certManagerClient) Images() ([]string, error) {
+	log.Info("Installing cert-manager")
 
-	// installs the web-hook
+	// Install all cert-manager manifests
```
Question. Should we start to add a version label to each component so we can properly plan for cert-manager upgrades?

munnerz replied:
This definitely seems sensible - but we will need to store that version number somewhere, and remember to update it if the YAML asset changes. I wonder if we can have an automated test for this..? Maybe even just take a hash of the YAML file and have a variable: `versionNumber` which is a tuple of the version number and the hash of that file?

That way if you change the file, the hash will no longer match, forcing people to look at that area of code?

Perhaps you have other instances of this somewhere and have an existing pattern to solve it? 😄

munnerz replied:
Done in https://github.com/kubernetes-sigs/cluster-api/pull/3313/commits/4aa658b1ff158452d4987c1727985357c33a1fe2 - keen to hear your comments on the technique for keeping the version number up to date.

fabriziopandini replied:
I personally like the idea of the hash, so we can manage adjustment to the manifest without bumping a release.

fabriziopandini replied:
I personally like the idea of the hash, so we can manage adjustment to the manifest without bumping a release.

fabriziopandini commented on cmd/clusterctl/config/assets/cert-manager.yaml:
```diff
@@ -1,26 +1,42 @@
-# This is manually edited cert-manager v0.11.0 release.
```
@wfernandes could you kindly double check we are not introducing regressions here?

munnerz replied:
I took a look over the changes you'd made, and all of those changes are now part of the upstream release and should no longer be required (although if there were any more edits beyond what's documented in this comment, I cannot be sure).

wfernandes replied:
Yeah...and we got tests to prove it 😉 YAY!

fabriziopandini commented on cmd/clusterctl/client/init.go:
```diff
@@ -105,7 +105,7 @@ func (c *clusterctlClient) Init(options InitOptions) ([]Components, error) {
 
 	// Before installing the providers, ensure the cert-manager Webhook is in place.
-	if err := cluster.CertManager().EnsureWebhook(); err != nil {
+	if err := cluster.CertManager().EnsureCertManager(); err != nil {
```
nit
```suggestion
	if err := cluster.CertManager().EnsureInstalled(); err != nil {
```
or something else to avoid to repeat CertManager

fabriziopandini commented on cmd/clusterctl/client/cluster/cert_manager.go:
```diff
@@ -17,23 +17,36 @@ limitations under the License.
+	// the version number marker here is _also_ updated.
+	// If the cert-manager.yaml asset is modified, this line **MUST** be updated
+	// accordingly else future upgrades of the cert-manager component will not
+	// be possible, as there'll be no record of the version installed.
```
might be adding a comment on how to generate the hash?

fabriziopandini commented on cmd/clusterctl/client/cluster/cert_manager.go:
```diff
@@ -17,23 +17,36 @@ limitations under the License.
+	// If the cert-manager.yaml asset is modified, this line **MUST** be updated
+	// accordingly else future upgrades of the cert-manager component will not
+	// be possible, as there'll be no record of the version installed.
+	embeddedCertManagerManifestVersion = map[string]string{"v0.15.2": "56bd9851567be27698114a73625594a9"}
```
Wondering if we should split this into two labels, so we can rely on a well know label name that will simplify the implementation of the upgrade logic eg.
```
clusterctl.cluster.x-k8s.io/version: "v0.15.2"
clusterctl.cluster.x-k8s.io/hash: "56bd9851567be27698114a73625594a9"
```

munnerz commented:
Addressed your feedback - should be good for another look!

The only other thing I'm concerned about now is the limitations/requirements about label values - from what I can tell, every possible character we are using right now _is_ valid/part of the character set permitted for labels. But it may be better to use annotations for some of this data if indexing/selecting based on these values is not required :)

munnerz commented:
Seeing:

```
cmd/clusterctl/client/cluster/cert_manager_test.go:41:34: G401: Use of weak cryptographic primitive (gosec)
	actualHash := fmt.Sprintf("%x", md5.Sum(yaml))
```

Before we go ahead and merge this, do we want to stick with MD5 for this hashing? I don't think we are so concerned around the cryptographic properties of the hashing function, but it'll be a lot easier to get it right now than it would be to change it later 😄

vincepri commented:
Let's just use sha256?

munnerz commented:
> Let's just use sha256?

Done ✅

vincepri commented:
Seems like we might have to limit the hash to be in an annotation, IIRC usually can be limited to 10 chars

wfernandes commented on cmd/clusterctl/client/cluster/cert_manager.go:
```diff
@@ -72,15 +88,8 @@ func newCertMangerClient(configClient config.Client, proxy Proxy, pollImmediateW
 }
 
 // Images return the list of images required for installing the cert-manager.
+// TODO: should this method return an empty slice if cert-manager is already installed?
```
This method is used when we run `clusterctl init --list-images`. So I don't believe that it should ever return an empty slice.

wfernandes commented on cmd/clusterctl/client/cluster/cert_manager.go:
```diff
@@ -220,86 +228,90 @@ func (cm *certManagerClient) createObj(o unstructured.Unstructured) error {
+	}
+	for i := range testObjs {
+		obj := testObjs[i]
+		if err := cm.deleteObj(obj); err != nil {
+			log.V(5).Info("Failed to delete cert-manager test resource", logf.UnstructuredToValues(obj)...)
+			return err
```
I feel if there may be some interesting behavior here. If we delete the namespace `cert-manager-test` and then delete the certificate/issuer could we run into a chance of getting a NotFound error since deleting the namespace would delete the stuff inside it?

I haven't run into it while testing manually...but I'm wondering if we should ignore NotFound errors here. WDYT?

fabriziopandini replied:
Might be we can simplify and simply delete the ns...

munnerz replied:
I've adjusted this to tolerate NotFound errors instead for now, given there is no specific namespace creation logic this seems simpler than having to inspect the manifests and infer what namespace to delete. This are can be improved as you've noted above to randomise the namespace name, which I think will make deletion a little simpler then too 😄

vincepri commented:
Hey folks, now that we've released v0.3.7 I was wondering about the status of this PR, any particular updates or action items?

fabriziopandini commented:
I'm +1 to rally and get this merged. 
My comment are already addressed and I have a WIP PR ready for covering the upgrade of existing clusters as soon as this one merges.
But I see some comments from @wfernandes still to be addressed

wfernandes commented:
If we are looking to get this merged in sooner rather than later, we can create separate issues and address some of the comments above. My comments aren't necessarily blockers.
However, we still need to fix the cause of the failing e2e test.

vincepri commented:
/milestone v0.3.9

fabriziopandini commented:
@munnerz is it ok for you trying to get this merged in this iteration?

vincepri commented:
Thank you @munnerz for finishing this up!

Changes LGTM, can we squash commits?

/assign @fabriziopandini @wfernandes

vincepri commented:
/approve
/hold
Feel free to unhold after squash

wfernandes commented:
I'll do another pass on this now. Thanks @munnerz!

wfernandes commented on cmd/clusterctl/client/cluster/cert_manager_test.go:
```diff
@@ -25,10 +27,24 @@ import (
 	"sigs.k8s.io/cluster-api/cmd/clusterctl/internal/test"
 )
 
+func Test_VersionMarkerUpToDate(t *testing.T) {
```
nit: we can use `g:=NewWithT(t)` and use gomega matchers here.
This is not a blocker though.

fabriziopandini commented:
Thanks @munnerz! looks great!
After squash, this is ready to go!!!

munnerz commented:
Done! Squashed and reworded the commit message to be summarise the changes.

munnerz commented:
For visibility on the apidiff failure, as discussed:

```
sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster
  Incompatible changes:
  - CertManagerClient.EnsureInstalled: added
  - CertManagerClient.EnsureWebhook: removed
```

fabriziopandini commented:
/lgtm
/approve
Thanks!

fabriziopandini commented:
/hold cancel
